### PR TITLE
Return error when distinct is used on unknown columns (#2867).

### LINKED ISSUE #2867: Feature request: return error when errant column used in distinct()
Can dplyr error out when one identifies a variable that does not exist in the data frame?

```r
## No error returned despite aa not appearing in the complimentary R data frame named "cars"
df <- cars %>% distinct(aa)

```

### PR DESCRIPTION
**Pull request by foo-bar-baz-qux**
This fixes #2867. Passes tests for distinct, also added a new test for this issue.

### CODE REVIEW

File: R/distinct.R [hunk 1]
```diff
@@ -76,7 +76,13 @@ distinct_vars <- function(.data, vars, group_vars = character(), .keep_all = FAL
 
   # Once we've done the mutate, we no longer need lazy objects, and
   # can instead just use their names
-  vars <- intersect(names(.data), c(names(vars), group_vars))
+  grouped_and_selected_vars <- unique(c(names(vars), group_vars))
+  vars <- intersect(names(.data), grouped_and_selected_vars)
+
+  if(length(vars) != length(grouped_and_selected_vars)) {
+    missing_vars <- setdiff(grouped_and_selected_vars, names(.data))
+    stop('Columns ', paste(missing_vars, collapse = ', '), ' are unknown')
```
Comment (krlmlr) [inline R/distinct.R]: Could you please use `bad_cols()`?

Comment (foo-bar-baz-qux) [inline R/distinct.R]: Thanks for reviewing, I've just made those changes.

File: R/distinct.R [hunk 2]
```diff
@@ -76,7 +76,13 @@ distinct_vars <- function(.data, vars, group_vars = character(), .keep_all = FAL
 
   # Once we've done the mutate, we no longer need lazy objects, and
   # can instead just use their names
-  vars <- intersect(names(.data), c(names(vars), group_vars))
+  grouped_and_selected_vars <- unique(c(names(vars), group_vars))
+  vars <- intersect(names(.data), grouped_and_selected_vars)
+
+  if(length(vars) != length(grouped_and_selected_vars)) {
+    missing_vars <- setdiff(grouped_and_selected_vars, names(.data))
```
Comment (hadley) [inline R/distinct.R]: I'd rather alway compute this and check it's length for the error

File: R/distinct.R [hunk 3]
```diff
@@ -76,7 +76,13 @@ distinct_vars <- function(.data, vars, group_vars = character(), .keep_all = FAL
 
   # Once we've done the mutate, we no longer need lazy objects, and
   # can instead just use their names
-  vars <- intersect(names(.data), c(names(vars), group_vars))
+  grouped_and_selected_vars <- unique(c(names(vars), group_vars))
```
Comment (hadley) [inline R/distinct.R]: `output_vars` might be a better/shorter name

File: R/distinct.R (hunk 2 above)
Comment (foo-bar-baz-qux) [inline R/distinct.R]: Makes sense, I've just pushed a new commit. I've also simplified this to a single set operation and clearer logic, plus added a test case to make sure new variables created on-the-fly through `distinct()` don't throw an error.

File: NEWS.md [hunk 4]
```diff
@@ -1,5 +1,7 @@
 # dplyr 0.7.1.9000
 
+* `distinct()` now throws an error when used on unknown columns (#2867).
```
Comment (hadley) [inline NEWS.md]: Can you please add your user name to the parenthetical?

File: R/distinct.R [hunk 5]
```diff
@@ -76,7 +76,12 @@ distinct_vars <- function(.data, vars, group_vars = character(), .keep_all = FAL
 
   # Once we've done the mutate, we no longer need lazy objects, and
   # can instead just use their names
-  vars <- intersect(names(.data), c(names(vars), group_vars))
+  vars <- unique(c(names(vars), group_vars))
+  missing_vars <- setdiff(vars, names(.data))
+
+  if(length(missing_vars) > 0) {
```
Comment (hadley) [inline R/distinct.R]: Missing space to left of `(`

File: tests/testthat/test-distinct.R [hunk 6]
```diff
@@ -67,3 +67,18 @@ test_that("empty grouped distinct equivalent to empty ungrouped", {
...
+test_that("distinct returns an error when selecting an unknown column", {
+  df <- tibble(g = c(1, 2), x = c(1, 2))
+
+  expect_error(distinct(df, aa))
+})
+
+test_that("distinct on a new, mutated variable is equivalent to mutate followed by distinct", {
+  df <- tibble(g = c(1, 2), x = c(1, 2))
+
+  df1 <- df %>% distinct(aa = g*2)
```
Comment (hadley) [inline tests/testthat/test-distinct.R]: Missing space around `*`

Comment (foo-bar-baz-qux) [inline tests/testthat/test-distinct.R]: Updated with those changes, cheers.

**Comment by hadley** [comment]
Thanks!

**Pull request merged**